<xacro:macro name="knuckle_link" params="prefix fingerprefix">
    <link name="${prefix}${fingerprefix}${knuckle_link_name}">
        <visual>
            <origin xyz="${knuckle_visual_xyz}" rpy="${knuckle_visual_rpy}" />
            <geometry>
                <mesh filename="${knuckle_visual_mesh_file}" scale="${knuckle_visual_mesh_scale}" />
            </geometry>
            <material name="${knuckle_material_name}">
                <color rgba="${knuckle_material_color}" />
            </material>
        </visual>
        <collision>
            <origin xyz="${knuckle_collision_xyz}" rpy="${knuckle_collision_rpy}" />
            <geometry>
                <mesh filename="${knuckle_collision_mesh_file}" scale="${knuckle_collision_mesh_scale}" />
            </geometry>
        </collision>
        <inertial>
            <origin xyz="${knuckle_inertial_xyz}" rpy="${knuckle_inertial_rpy}" />
            <mass value="${knuckle_mass}" />
            <inertia ixx="${knuckle_ixx}" ixy="${knuckle_ixy}" ixz="${knuckle_ixz}" iyy="${knuckle_iyy}" iyz="${knuckle_iyz}" izz="${knuckle_izz}" />
        </inertial>
    </link>
</xacro:macro>

<xacro:macro name="bar_link" params="prefix fingerprefix">
    <link name="${prefix}${fingerprefix}${bar_link_name}">
        <visual>
            <origin xyz="${bar_visual_xyz}" rpy="${bar_visual_rpy}" />
            <geometry>
                <mesh filename="${bar_visual_mesh_file}" scale="${bar_visual_mesh_scale}" />
            </geometry>
            <material name="${bar_material_name}">
                <color rgba="${bar_material_color}" />
            </material>
        </visual>
        <collision>
            <origin xyz="${bar_collision_xyz}" rpy="${bar_collision_rpy}" />
            <geometry>
                <mesh filename="${bar_collision_mesh_file}" scale="${bar_collision_mesh_scale}" />
            </geometry>
        </collision>
        <inertial>
            <origin xyz="${bar_inertial_xyz}" rpy="${bar_inertial_rpy}" />
            <mass value="${bar_mass}" />
            <inertia ixx="${bar_ixx}" ixy="${bar_ixy}" ixz="${bar_ixz}" iyy="${bar_iyy}" iyz="${bar_iyz}" izz="${bar_izz}" />
        </inertial>
    </link>
</xacro:macro>

<xacro:macro name="proximal_phalanx_link" params="prefix fingerprefix">
    <link name="${prefix}${fingerprefix}${proximal_phalanx_link_name}">
        <visual>
            <origin xyz="${proximal_phalanx_visual_xyz}" rpy="${proximal_phalanx_visual_rpy}" />
            <geometry>
                <mesh filename="${proximal_phalanx_visual_mesh_file}" scale="${proximal_phalanx_visual_mesh_scale}" />
            </geometry>
            <material name="${proximal_phalanx_material_name}">
                <color rgba="${proximal_phalanx_material_color}" />
            </material>
        </visual>
        <collision>
            <origin xyz="${proximal_phalanx_collision_xyz}" rpy="${proximal_phalanx_collision_rpy}" />
            <geometry>
                <mesh filename="${proximal_phalanx_collision_mesh_file}" scale="${proximal_phalanx_collision_mesh_scale}" />
            </geometry>
        </collision>
        <inertial>
            <origin xyz="${proximal_phalanx_inertial_xyz}" rpy="${proximal_phalanx_inertial_rpy}" />
            <mass value="${proximal_phalanx_mass}" />
            <inertia ixx="${proximal_phalanx_ixx}" ixy="${proximal_phalanx_ixy}" ixz="${proximal_phalanx_ixz}" iyy="${proximal_phalanx_iyy}" iyz="${proximal_phalanx_iyz}" izz="${proximal_phalanx_izz}" />
        </inertial>
    </link>
</xacro:macro>

<xacro:macro name="distal_phalanx_link" params="prefix fingerprefix">
    <link name="${prefix}${fingerprefix}${distal_phalanx_link_name}">
        <visual>
            <origin xyz="${distal_phalanx_visual_xyz}" rpy="${distal_phalanx_visual_rpy}" />
            <geometry>
                <mesh filename="${distal_phalanx_visual_mesh_file}" scale="${distal_phalanx_visual_mesh_scale}" />
            </geometry>
            <material name="${distal_phalanx_material_name}">
                <color rgba="${distal_phalanx_material_color}" />
            </material>
        </visual>
        <collision>
            <origin xyz="${distal_phalanx_collision_xyz}" rpy="${distal_phalanx_collision_rpy}" />
            <geometry>
                <mesh filename="${distal_phalanx_collision_mesh_file}" scale="${distal_phalanx_collision_mesh_scale}" />
            </geometry>
        </collision>
        <inertial>
            <origin xyz="${distal_phalanx_inertial_xyz}" rpy="${distal_phalanx_inertial_rpy}" />
            <mass value="${distal_phalanx_mass}" />
            <inertia ixx="${distal_phalanx_ixx}" ixy="${distal_phalanx_ixy}" ixz="${distal_phalanx_ixz}" iyy="${distal_phalanx_iyy}" iyz="${distal_phalanx_iyz}" izz="${distal_phalanx_izz}" />
        </inertial>
    </link>
</xacro:macro>

<xacro:macro name="finger_tip_link" params="prefix fingerprefix">
    <link name="${prefix}${fingerprefix}${finger_tip_link_name}">
        <visual>
            <origin xyz="${finger_tip_visual_xyz}" rpy="${finger_tip_visual_rpy}" />
            <geometry>
                <box size="${finger_tip_box_dimension}" />
            </geometry>
            <material name="${finger_tip_visual_material_name}">
                <color rgba="${finger_tip_visual_material_color}" />
            </material>
        </visual>
        <collision>
            <origin xyz="${finger_tip_collision_xyz}" rpy="${finger_tip_collision_rpy}" />
            <geometry>
                <box size="${finger_tip_box_dimension}" />
            </geometry>
            <material name="${finger_tip_collision_material_name}">
                <color rgba="${finger_tip_collision_material_color}" />
            </material>
        </collision>
    </link>
</xacro:macro>

<xacro:knuckle_link prefix="${prefix}" fingerprefix="${fingerprefix}" />
<xacro:bar_link prefix="${prefix}" fingerprefix="${fingerprefix}" />
<xacro:proximal_phalanx_link prefix="${prefix}" fingerprefix="${fingerprefix}" />
<xacro:distal_phalanx_link prefix="${prefix}" fingerprefix="${fingerprefix}" />
